move RoutedDecoder Dataset to DataPipe

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51709 move GroupByFilename Dataset to DataPipe
* **#51704 move RoutedDecoder Dataset to DataPipe**

This is to make RoutedDecoder a DataPipe, currently using Nvidia's decode, we will refactor this later. We need this for now to run basic benchmark.

Differential Revision: [D26245910](https://our.internmc.facebook.com/intern/diff/D26245910)

Commit:
move RoutedDecoder Dataset to DataPipe

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/utils/decoder.py:
```diff
@@ -0,0 +1,270 @@
+                return result
+        return data
+
+    def decode(self, datapipe):
+        result = {}
+        # single data tuple(pathname, data stream)
+        if isinstance(datapipe, tuple):
+            datapipe = [datapipe]
+
+        if datapipe is not None:
+            for k, v in datapipe:
+                # TODO: xinyu, figure out why Nvidia do this?
+                if k[0] == "_":
+                    if isinstance(v, bytes):
+                        v = v.decode("utf-8")
+                        result[k] = v
+                        continue
+                result[k] = self.decode1(k, v)
+        return result
+
+    def __call__(self, datapipe):
+        return self.decode(datapipe)
```
Can we change `datapipe` to `data`, this name makes the logic confusing, as the input is actually one element for each iteration rather than the datapipe itself.
Besides, is there any benefit using dict as result?

glaringlee replied:
Will change datapipe to data. 
It is not clear whether dict has benefit at the moment, I am not sure whether we have any case that we need to decode a batch of image at once and need to locate a specific record within a dict. Let's keep this, we can change the code later to meet our needs.

ejguan reviewed (approved):
LGTM with some nit comments below.

Commit:
Update on "move RoutedDecoder Dataset to DataPipe"


This is to make RoutedDecoder a DataPipe, currently using Nvidia's decode, we will refactor this later. We need this for now to run basic benchmark.

Differential Revision: [D26245910](https://our.internmc.facebook.com/intern/diff/D26245910)

[ghstack-poisoned]

Commit:
Update on "move RoutedDecoder Dataset to DataPipe"


This is to make RoutedDecoder a DataPipe, currently using Nvidia's decode, we will refactor this later. We need this for now to run basic benchmark.

Differential Revision: [D26245910](https://our.internmc.facebook.com/intern/diff/D26245910)

[ghstack-poisoned]

Commit:
Update on "move RoutedDecoder Dataset to DataPipe"


This is to make RoutedDecoder a DataPipe, currently using Nvidia's decode, we will refactor this later. We need this for now to run basic benchmark.

Differential Revision: [D26245910](https://our.internmc.facebook.com/intern/diff/D26245910)

[ghstack-poisoned]
